src/main/telemetry/esc_telemetry.c

```diff
@@ -0,0 +1,319 @@
+    }
+}
+
+uint8_t escTelemetryFrameStatus(void)
+{
+    uint8_t frameStatus = ESC_TLM_FRAME_PENDING;
+    uint16_t chksum, tlmsum;
+
+    if (!tlmFrameDone) {
+        return frameStatus;
+    }
+
+    tlmFrameDone = false;
+
+    // Get CRC8 checksum
+    chksum = get_crc8(tlm, ESC_TLM_BUFFSIZE - 1);
+    tlmsum = tlm[ESC_TLM_BUFFSIZE - 1];     // last byte contains CRC value
+
+    if (chksum == tlmsum) {
+        escTelemetryData[escTelemetryMotor].skipped = false;
+        escTelemetryData[escTelemetryMotor].temperature = tlm[0];
+        escTelemetryData[escTelemetryMotor].voltage = tlm[1] << 8 | tlm[2];
+        escTelemetryData[escTelemetryMotor].current = tlm[3] << 8 | tlm[4];
+        escTelemetryData[escTelemetryMotor].consumption = tlm[5] << 8 | tlm[6];
+        escTelemetryData[escTelemetryMotor].rpm = tlm[7] << 8 | tlm[8];
+
+        frameStatus = ESC_TLM_FRAME_COMPLETE;
+    }
+
+    return frameStatus;
+}
+
+void escTelemetryProcess(uint32_t currentTime)
+{
+    uint32_t currentTimeMs = currentTime / 1000;
+
+    if (!escTelemetryEnabled) {
+        return;
+    }
+
+    // Wait period of time before requesting telemetry (let the system boot first)
+    if (millis() < ESC_BOOTTIME) {
+        return;
+    }
+    else if (escTelemetryTriggerState == ESC_TLM_TRIGGER_WAIT) {
+        // Ready for starting requesting telemetry
+        escTelemetryTriggerState = ESC_TLM_TRIGGER_READY;
+        escTelemetryMotor = 0;
+        escTriggerTimestamp = currentTimeMs;
+        escTriggerLastTimestamp = escTriggerTimestamp;
+    }
+    else if (escTelemetryTriggerState == ESC_TLM_TRIGGER_READY) {
+        if (debugMode == DEBUG_ESC_TELEMETRY) debug[0] = escTelemetryMotor+1;
+
+        motorDmaOutput_t * const motor = getMotorDmaOutput(escTelemetryMotor);
+        motor->requestTelemetry = true;
+        escTelemetryTriggerState = ESC_TLM_TRIGGER_PENDING;
+    }
+
+    if (escTriggerTimestamp + ESC_REQUEST_TIMEOUT < currentTimeMs) {
+        // ESC did not repond in time, retry
+        timeoutRetryCount++;
+        escTriggerTimestamp = currentTimeMs;
+        escTelemetryTriggerState = ESC_TLM_TRIGGER_READY;
+
+        if (timeoutRetryCount == 4) {
+            // Not responding after 3 times, skip motor
+            escTelemetryData[escTelemetryMotor].skipped = true;
+            selectNextMotor();
+        }
+
+        if (debugMode == DEBUG_ESC_TELEMETRY) debug[1]++;
+    }
+
+    // Get received frame status
+    uint8_t state = escTelemetryFrameStatus();
+
+    if (state == ESC_TLM_FRAME_COMPLETE) {
+        // Wait until all ESCs are processed
+        if (escTelemetryMotor == enabledMotorCount-1) {
+            escCurrent = 0;
+            escConsumption = 0;
+            escVbat = 0;
+
+            for (int i = 0; i < enabledMotorCount; i++) {
+                if (!escTelemetryData[i].skipped) {
+                    escVbat =  i > 0 ? ((escVbat + escTelemetryData[i].voltage) / 2) : escTelemetryData[i].voltage;
```

Comment:
This is not actually calculating the average, since the weighting is not equal.

It's probably better anyway if instead of calculating the average over all voltages we calculate the average over the 2 voltages that are closest, and consider the other ones to be outliers.